### PR TITLE
8 structures

### PR DESCRIPTION
**Pull request by Matte02**
Added structures to the game and can be placed by pressing left click.

### CODE REVIEW

File: src/main/java/com/thebois/models/world/World.java [hunk 1]
```diff
@@ -50,4 +54,40 @@ public Object find() {
         return copy;
     }
 
+    /**
+     * Returns the structures in an arraylist as the interface IStructures.
+     *
+     * @return The list to be returned.
+     */
+    public ArrayList<IStructure> getStructures() {
+        return structures;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/world/World.java:63]: Not a deep copy.

File: src/main/java/com/thebois/models/world/World.java [hunk 2]
```diff
@@ -3,13 +3,18 @@
...
+import com.thebois.models.world.structures.Structure;
+import com.thebois.models.world.structures.StructureType;
 
 /**
  * World creates a matrix and keeps track of all the structures and resources in the game world.
  */
 public class World implements IFinder {
 
     private ITerrain[][] terrainMatrix;
+    private ArrayList<IStructure> structures = new ArrayList<>();
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/world/World.java:17]: Think we sould stay consisten how we represent the world in the model. We will also use the matrix for pathfinding, so in the world we should have a `structureMatrix` and in `.getStructures` we convert it to an ArrayList (and deepcopy the elements) and return that.

File: src/main/java/com/thebois/models/Position.java [hunk 3]
```diff
@@ -42,6 +44,23 @@ public void setPosY(final float posY) {
         this.posY = posY;
     }
 
+    @Override
+    public boolean equals(final Object o) {
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/Position.java:48]: I remember we had a discussion if this equals method actually returned the correct value. Have we made sure it does?

File: src/main/java/com/thebois/models/world/structures/Structure.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
+package com.thebois.models.world.structures;
+
+import java.util.Objects;
+
+import com.thebois.models.Position;
+
+/**
+ * Base structure for the game.
+ */
+public class Structure implements IStructure {
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/world/structures/Structure.java:10]: We will never have just a structure placed in the world so this class could be made abstract imo. And have a House class implement it.

File: src/main/java/com/thebois/views/StructureView.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * Gets the tile size of the world.
+     *
+     * @return the tile size as a float.
+     */
+    public float getTileSize() {
+        return tileSize;
+    }
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/StructureView.java:45]: See comment related in controller.

File: src/main/java/com/thebois/views/WorldView.java [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
     private final Color grassColor = new Color(0.005f, 0.196f, 0.107f, 1);
     private ArrayList<ITerrain> terrainTiles = new ArrayList<>();
-    private float tileSize;
+    private final float tileSize;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/WorldView.java:17]: Have you tested that it scales correctly after making this final?

File: src/test/java/com/thebois/models/world/StructureTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Test
+    public void getTypeTest() {
+
+        // Arrange
+        final Position position = new Position(123, 456);
+
+        // Act
+        final Structure structure = new Structure(position, StructureType.HOUSE);
```
Comment (jonathan-lindqvist) [inline src/test/java/com/thebois/models/world/StructureTest.java:34]: Not sure this is a part of Act since it just sets up a structure, we don't act on the object.

File: src/test/java/com/thebois/models/world/WorldTests.java [hunk 8]
```diff
@@ -43,6 +46,35 @@ public void worldFind() {
...
+        // Arrange
+        final World world = new World(25);
+
+        // Act
+        final Boolean isEmpty = world.getStructures().isEmpty();
+        final Boolean isStructureBuilt = world.createStructure(new Position(0, 0));
+        final int numberOfStructures = world.getStructures().size();
+
+        // Assert
+        assertThat(isEmpty).isTrue();
```
Comment (jonathan-lindqvist) [inline src/test/java/com/thebois/models/world/WorldTests.java:73]: Not sure if this should be part of `createStructureTest` since it only tests if we can create an empty structure world. Maybe make it a seperate test.

**Review by jonathan-lindqvist — CHANGES_REQUESTED** [review]
Nicely done overall, just some specific implementations that should/could be tweaked.

And just a reminder to start the commit with a capital letter, e.g. "Add ..." vs "add ...", and to summarize the pull request in the description. But absolutely no big deal, just nice to do to get into the habit for the future and if we quickly want to go back and look at how we did something.

File: src/main/java/com/thebois/views/WorldView.java (hunk 6 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/WorldView.java:17]: Yes, since we actually never change the size on the tiles. When we resize the window it scales everything inside the viewport which means the tile size remains the same. And if you check everywhere we use tile size we never actually update it.

File: src/main/java/com/thebois/ColonyManagement.java [hunk 9]
```diff
@@ -26,9 +31,11 @@
     private World world;
     // Views
     private WorldView worldView;
+    private StructureView structureView;
     // Screens
     private GameScreen gameScreen;
     private TerrainController terrainController;
+    private StructureController structureController;
```
[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/ColonyManagement.java:38]: Might have been me who moved it but will move them down to controller commet.

File: src/main/java/com/thebois/models/Position.java (hunk 3 above)
Comment (Matte02) [reply src/main/java/com/thebois/models/Position.java:48]: I believe it should since 2 positions are equal if they have the same x and y value.

File: src/main/java/com/thebois/models/world/World.java (hunk 2 above)
Comment (Matte02) [reply src/main/java/com/thebois/models/world/World.java:17]: I thought about it when writing the code, but it felt wrong to have a matrix that would be mostly empty. We can discuss this tomorrow during the meeting.

File: src/main/java/com/thebois/models/world/World.java (hunk 1 above)
Comment (Matte02) [reply src/main/java/com/thebois/models/world/World.java:63]: Oh right, will fix this.

File: src/test/java/com/thebois/models/world/StructureTest.java (hunk 7 above)
Comment (Matte02) [reply src/test/java/com/thebois/models/world/StructureTest.java:34]: You're probably right, I'll just remove the act commet.

File: src/main/java/com/thebois/ColonyManagement.java [hunk 10]
```diff
@@ -47,16 +54,24 @@ public void create() {
 
         // Views
         worldView = new WorldView(tileSize);
+        structureView = new StructureView(tileSize);
+
+        // Arrange Views for gameScreen
         final ArrayList<IView> views = new ArrayList<>();
         views.add(worldView);
+        views.add(structureView);
```
[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/ColonyManagement.java:62]: Sounds good!

File: src/main/java/com/thebois/controllers/StructureController.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+     * @param world         The world in which the structures exists.
+     * @param structureView A view where the structures will be displayed.
+     * @param projector     Projector used for converting screen coordinates to world coordinates.
+     */
+    public StructureController(final World world,
+                               final StructureView structureView,
+                               final IProjector projector) {
+        this.world = world;
+        this.structureView = structureView;
+        tileSize = structureView.getTileSize();
```
[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/controllers/StructureController.java:32]: It's probably better to do that.

File: src/main/java/com/thebois/models/world/structures/Structure.java (hunk 4 above)
Comment (Matte02) [reply src/main/java/com/thebois/models/world/structures/Structure.java:10]: Since there is no difference between a house and a structure it doesn't really matter, but it would make sense to make structure abstract for the future. Will fix.

File: src/main/java/com/thebois/views/StructureView.java (hunk 5 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/StructureView.java:45]: Yes, will remove this.

File: src/test/java/com/thebois/models/world/WorldTests.java (hunk 8 above)
Comment (Matte02) [reply src/test/java/com/thebois/models/world/WorldTests.java:73]: Wanted to make sure the world is empty and not already containing a structure because if there was and createStructure was made incorrectly and return true this test would pass even though createStructure method doesn't work. Should I write the test in another way?

File: src/main/java/com/thebois/views/GameScreen.java [hunk 12]
```diff
@@ -47,6 +47,10 @@ public GameScreen(final ShapeRenderer batch,
         this.views = views;
     }
 
+    public IProjector getProjector() {
+        return new ViewportWrapper(viewport);
+    }
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/GameScreen.java:52]: No javadoc?

File: src/main/java/com/thebois/views/IProjector.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
+package com.thebois.views;
+
+import com.badlogic.gdx.math.Vector2;
+
+/**
+ * Interface for a Projector.
+ */
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/IProjector.java:7]: You can do better than this. 😛

File: src/main/java/com/thebois/views/IProjector.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+public interface IProjector {
+
+    /**
+     * Unprojects screen coordiantes to world coordinates.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/IProjector.java:11]: *coordinates*

File: src/main/java/com/thebois/views/IProjector.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+     *
+     * @param screenCordX X coordinate of the screen.
+     * @param screenCordY Y coordinate of the screen.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/IProjector.java:14]: These are better described as the x- and y-coordinates *of a position in screen space*. **Not** the position *of the screen*.

File: src/main/java/com/thebois/views/IProjector.java [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+     *
+     * @return Returns a Vector2 with the world coordinates.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/IProjector.java:16]: "a Vector2 in world space" / "in world coordinates"

File: src/main/java/com/thebois/views/IProjector.java [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+     */
+    Vector2 unProject(int screenCordX, int screenCordY);
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/IProjector.java:18]: *unproject*

File: src/main/java/com/thebois/views/StructureView.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+import com.thebois.models.world.structures.IStructure;
+
+/**
+ * Responsible for displaying structures.
+ */
+public final class StructureView implements IView {
+
+    private final float tileSize;
+    private ArrayList<IStructure> structures = new ArrayList<>();
+    private final Color brown = new Color(0.4f, 0.2f, 0f, 1);
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/StructureView.java:17]: A better name would be `houseColor`

File: src/main/java/com/thebois/views/StructureView.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * Creates an instance of a Structure view.
+     *
+     * @param tileSize The tile size of the world in order to render structures in the right size.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/StructureView.java:22]: Not sure if the reason why it is needed should be here. I think "The size of a single world tile in screen space" would fit better.

Comment (martinjonsson01) [reply src/main/java/com/thebois/views/StructureView.java:22]: (The reason I'm being so picky about wording is to reduce confusion in the future when we need to look back at and understand what we meant today.)

File: src/main/java/com/thebois/views/StructureView.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    public float getTileSize() {
+        return tileSize;
+    }
+
+    /**
+     * Updates the structures that should be displayed in the game.
+     *
+     * @param Structures The structures that are to be displayed.
+     */
+    public void updateStructureView(ArrayList<IStructure> Structures) {
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/StructureView.java:52]: The class is called `StructureView` so calling `StructureView.updateStructureView` is quite redundant. Better to just call it `update` since we will be calling `StructureView.update`.

Comment (martinjonsson01) [reply src/main/java/com/thebois/views/StructureView.java:52]: Oh, and the argument shouldn't be capitalized, it should be `structures`.

File: src/main/java/com/thebois/views/ViewportWrapper.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
+package com.thebois.views;
+
+import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.utils.viewport.Viewport;
+
+/**
+ * Wrapper class for viewports.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/ViewportWrapper.java:7]: Since the class only wraps a *single* viewport it's probably better to not pluralize it here.

File: src/test/java/com/thebois/models/world/StructureTest.java [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+import com.thebois.models.world.structures.StructureType;
+
+import static org.assertj.core.api.Assertions.*;
+
+public class StructureTest {
+
+    @Test
+    public void getPositionTest() {
+
+        // Arrange
```
Comment (martinjonsson01) [inline src/test/java/com/thebois/models/world/StructureTest.java:16]: Am I the only one who doesn't like this space?

File: src/test/java/com/thebois/models/world/StructureTest.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+import com.thebois.models.Position;
+import com.thebois.models.world.structures.Structure;
+import com.thebois.models.world.structures.StructureType;
+
+import static org.assertj.core.api.Assertions.*;
+
+public class StructureTest {
+
+    @Test
+    public void getPositionTest() {
```
Comment (martinjonsson01) [inline src/test/java/com/thebois/models/world/StructureTest.java:14]: Just some of my opinions regarding test naming:

* There's no point in having the word "test" in the name, since all the methods in the class are tests.
* The test name should describe *which* method is being tested (getPosition) and *what is being confirmed* (ReturnsPositionSetInConstructor) for a final name of `getPositionReturnsPositionSetInConstructor`.
* There is no such thing as a test with too long a name.

File: src/test/java/com/thebois/models/world/StructureTest.java [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+
+        // Act
+        final Position structurePosition = structure.getPosition();
+
+        // Assert
+        assertThat(structurePosition).isEqualTo(position);
+    }
+
+    @Test
+    public void getTypeTest() {
```
Comment (martinjonsson01) [inline src/test/java/com/thebois/models/world/StructureTest.java:28]: As an example, this name doesn't tell me anything about *what is being tested*, only *which* method is being tested.

File: src/test/java/com/thebois/models/world/StructureTest.java (hunk 7 above)
Comment (martinjonsson01) [reply src/test/java/com/thebois/models/world/StructureTest.java:34]: On the other hand, what is being tested here is that the constructor sets the value correctly. So in this case you could argue that *constructing* the object is the *acting*.

File: src/test/java/com/thebois/models/world/WorldTests.java [hunk 25]
```diff
@@ -43,6 +46,35 @@ public void worldFind() {
         assertThat(worldObject).isEqualTo(null);
     }
 
+    @Test
+    public void getStructuresTest() {
+        // Arrange
+        final World world = new World(25);
+
+        // Act
+        world.createStructure(new Position(123, 123));
```
Comment (martinjonsson01) [inline src/test/java/com/thebois/models/world/WorldTests.java:55]: This line is only for setting up the test. The line below, where `world.getStructures()` is called, is the acting.

File: src/test/java/com/thebois/models/world/WorldTests.java [hunk 26]
```diff
@@ -43,6 +46,35 @@ public void worldFind() {
...
+        final IStructure structure = world.getStructures().get(0);
+
+        // Assert
+        assertThat(structure.getPosition()).isEqualTo(new Position(123, 123));
```
Comment (martinjonsson01) [inline src/test/java/com/thebois/models/world/WorldTests.java:59]: To follow DRY, you could break out this `new Position(123, 123)` and the one on line 55 into a single constant that you set under `// Arrange`. This makes it easier to modify the test later on.

File: src/test/java/com/thebois/models/world/WorldTests.java (hunk 8 above)
Comment (martinjonsson01) [reply src/test/java/com/thebois/models/world/WorldTests.java:73]: You should have multiple tests:
`worldContainsNoStructuresFromStart()`
`numberOfStructuresGoesUpWhenStructureIsSuccessfullyPlaced()`
`numberOfStructuresDoesNotGoUpWhenStructureIsUnsuccessfullyPlaced()`

**Review by martinjonsson01 — CHANGES_REQUESTED** [review]
Looks good overall, just some minor tweaks here and there.

File: src/main/java/com/thebois/models/world/World.java (hunk 2 above)
Comment (Matte02) [reply src/main/java/com/thebois/models/world/World.java:17]: > Yeah, the pathfinding will need to be able to easily find neighbouring structures as well as terrain tiles. Same goes for resources whenever we implement them. They should all be matrices like `terrainMatrix` and returned in the same way that `getStructures` does.

Have I understood this corectly by saying: You want the structures to be stored in a Matrix like the tiles, but in the getter it should be return as a Arraylist(like it is now, but a deep copy instead)?

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:17]: Yes, exactly. Should be the same as for `ITerrain`.

File: src/main/java/com/thebois/models/world/structures/IStructure.java [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+ * Represents a Structure. E.g. a house.
+ */
+public interface IStructure extends ITile {
+
+    /**
+     * Returns the specific type of terrain.
+     *
+     * @return The terrain type.
+     */
+    StructureType getType();
```
[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/models/world/structures/IStructure.java:15]: Whoops, fixing it!

File: src/main/java/com/thebois/views/GameScreen.java (hunk 12 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/GameScreen.java:52]: Hmm, checkstyle isn't picking that up, which is really strange considering it's not in the suppression XML file. Will add it.

File: src/test/java/com/thebois/models/world/StructureTest.java (hunk 22 above)
Comment (Matte02) [reply src/test/java/com/thebois/models/world/StructureTest.java:16]: Space as in the new line between or space between // and Arrange?

File: src/main/java/com/thebois/controllers/StructureController.java [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+        structureView.updateStructureView(world.getStructures());
+        this.projector = projector;
```
[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/controllers/StructureController.java:34]: On it.

File: src/main/java/com/thebois/controllers/StructureController.java [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public boolean touchDown(final int x, final int y, final int pointer, final int button) {
+        if (button == LEFT_CLICK) {
+            final Vector2 worldSpaceCords = projector.unProject(x, y);
+            final int worldPosX = (int) (worldSpaceCords.x / tileSize);
+            final int worldPosY = (int) (worldSpaceCords.y / tileSize);
+
+            if (world.createStructure(new Position(worldPosX, worldPosY))) {
```
[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/controllers/StructureController.java:44]: Probably better, will change it to that.

File: src/main/java/com/thebois/views/GameScreen.java (hunk 12 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/GameScreen.java:52]: Just realized that getters and setters don't need javadoc. Nevermind. Maybe this one should be an exception though?

File: src/main/java/com/thebois/models/Position.java (hunk 3 above)
Comment (Matte02) [reply src/main/java/com/thebois/models/Position.java:48]: Thought for some reason we already had those tests, but apparently not. I'll add them.

File: src/main/java/com/thebois/views/IProjector.java (hunk 13 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/IProjector.java:7]: Should probably have waited with writing the javadoc, and not do it at the end of the day. Changing it to something better.

File: src/main/java/com/thebois/views/IProjector.java (hunk 17 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/IProjector.java:18]: Intellij's spelling checker complained but will change it and tell it to allow it.

File: src/main/java/com/thebois/views/StructureView.java (hunk 19 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/StructureView.java:22]: It's good that you are picky, hopefully, I learn and do better next time.

File: src/main/java/com/thebois/views/StructureView.java (hunk 20 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/StructureView.java:52]: Changing it.

File: src/main/java/com/thebois/views/ViewportWrapper.java (hunk 21 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/ViewportWrapper.java:7]: Makes sense.

File: src/test/java/com/thebois/models/world/StructureTest.java (hunk 23 above)
Comment (Matte02) [reply src/test/java/com/thebois/models/world/StructureTest.java:14]: Will take a look at the tests and add the missing ones and give them proper names.

File: src/main/java/com/thebois/views/StructureView.java (hunk 20 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/StructureView.java:52]: > Oh, and the argument shouldn't be capitalized, it should be `structures`.

@martinjonsson01 
image
I still feel like structures is a good argument name, opinion on this checkstyle error?

File: src/main/java/com/thebois/views/GameScreen.java (hunk 12 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/GameScreen.java:52]: I agree with you, I'm adding javadoc for it.

File: src/main/java/com/thebois/views/StructureView.java (hunk 20 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/StructureView.java:52]: Hmm, normally it's good that it warns about hiding fields as it's simple to miss and can cause confusing issues. In this case we have a method that is acting like a setter, but isn't named like one, so checkstyle doesn't allow it the exemption it does for setters.

I'd call it something like `newStructures` or `updatedStructures`. We did something similar in the update method for `ITerrain`.

File: src/test/java/com/thebois/models/world/StructureTest.java (hunk 22 above)
Comment (martinjonsson01) [reply src/test/java/com/thebois/models/world/StructureTest.java:16]: The newline, `// Arrange` is correct.

File: src/test/java/com/thebois/models/world/StructureTest.java (hunk 23 above)
Comment (martinjonsson01) [reply src/test/java/com/thebois/models/world/StructureTest.java:14]: Another thing - I think we've been calling all our test classes `xxxxTests` and not `xxxxTest` so far. Best to follow that as well.

File: src/main/java/com/thebois/models/world/structures/Structure.java [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * Creates a structure with a position.
+     *
+     * @param posX Position in X-axis
+     * @param posY Position in Y-axis
+     */
+    public Structure(int posX, int posY) {
+        this.position = new Position(posX, posY);
+    }
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/structures/Structure.java:34]: Just like is done in `Position`:
image

File: src/main/java/com/thebois/ColonyManagement.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/ColonyManagement.java:61]: > 
> 
> Isn't our current `WorldView` more of a `TerrainView`?
> 
> Both `TerrainView` and `StructureView` should be wrapped in a `WorldView` that renders the entirety of the world, including arranging the views for it.

That would be neat but I am referring to arranging of view does not really belong here, right now its OK to thave them here but soon we might have 5 views.

[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/ColonyManagement.java:61]: Yeah that's true, the problem still persists. We should discuss this tomorrow, I'll add it to the discussion points.

[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/ColonyManagement.java:61]: Haven't changed anything here, will do changes after we have discussed this tomorrow.

**Review by martinjonsson01 — APPROVED** [review]
Looks great, well done!

File: src/main/java/com/thebois/models/world/World.java [hunk 31]
```diff
@@ -31,7 +30,7 @@ public World(int worldSize) {
         }
 
         // Structures
-        structuresMatrix = new Structure[worldSize][worldSize];
+        structuresMatrix = new House[worldSize][worldSize];
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/models/world/World.java:33]: This is not a good way of instantiating the matrix. I can see why you used `House` as both `IStructure` and `AbstractStructure` can't be instantiated, but I think you can still create an `IStructure[worldSize][worldSize]`. Or am I wrong? 

If it is not possible, then we need to rethink this, because the matrix needs to be able to hold `Structure`s other than `House`s.

**Review by martinjonsson01 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thebois/views/IProjector.java [hunk 32]
```diff
@@ -0,0 +1,20 @@
...
+
+    /**
+     * Unprojects screen coordinates to world coordinates.
+     *
+     * @param screenCordX X coordinate of a position in screen space.
+     * @param screenCordY Y coordinate of a position in screen space.
+     *
+     * @return Returns a Vector2 in the world coordinates.
+     */
+    Vector2 unproject(int screenCordX, int screenCordY);
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/IProjector.java:18]: This is no big deal at all since it's clear anyway, but we talked about writing the whole name, so `screenCoordinateX`. Or did we mean just not to write for example `sCX`?

File: src/main/java/com/thebois/views/PositionAdapter.java [hunk 33]
```diff
@@ -24,4 +24,15 @@ public static Vector2 toVector(Position position) {
...
     }
 
+    /**
+     * Converts vector to custom position.
+     *
+     * @param vector2 The vector to convert from.
+     *
+     * @return The same vector in position form.
+     */
+    public static Position toPosition(Vector2 vector2) {
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/PositionAdapter.java:34]: Think it would be better with `vector2D`, since the current name seems to refer to the 2nd vector.

File: src/main/java/com/thebois/views/ViewportWrapper.java [hunk 34]
```diff
@@ -0,0 +1,27 @@
...
+     * Creates an instance of the viewport wrapper.
+     *
+     * @param viewport the viewport to be wrapped.
+     */
+    protected ViewportWrapper(final Viewport viewport) {
+        this.viewport = viewport;
+    }
+
+    @Override
+    public Vector2 unproject(final int screenCordX, final int screenCordY) {
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/ViewportWrapper.java:23]: See the related comment in IProjector

File: src/test/java/com/thebois/models/world/WorldTests.java (hunk 8 above)
Comment (jonathan-lindqvist) [reply src/test/java/com/thebois/models/world/WorldTests.java:104]: Now that we have a test for this I think we can skip it here, right?

**Review by jonathan-lindqvist — CHANGES_REQUESTED** [review]
Some small discussion items, when resolved this PR should be able to merge! 👍

File: src/main/java/com/thebois/views/IProjector.java (hunk 32 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/IProjector.java:18]: Probably best to not have any exceptions, easier to remember that way. Writing out coordinate here is better.

File: src/main/java/com/thebois/views/PositionAdapter.java (hunk 33 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/PositionAdapter.java:34]: Or just `vector` since the dimensionality of it is apparent from its type.

File: src/test/java/com/thebois/models/world/WorldTests.java (hunk 8 above)
Comment (martinjonsson01) [reply src/test/java/com/thebois/models/world/WorldTests.java:104]: ^Yeah it's a bit redundant here.

File: src/main/java/com/thebois/models/world/World.java (hunk 31 above)
Comment (Matte02) [reply src/main/java/com/thebois/models/world/World.java:33]: IStructure probably works, I was a bit tired when refactoring, will take a look at this tomorrow.

File: src/main/java/com/thebois/views/IProjector.java (hunk 32 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/IProjector.java:18]: On my todo list, will do it tomorrow

File: src/main/java/com/thebois/views/PositionAdapter.java (hunk 33 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/PositionAdapter.java:34]: Vector it is.

File: src/test/java/com/thebois/models/world/WorldTests.java (hunk 8 above)
Comment (Matte02) [reply src/test/java/com/thebois/models/world/WorldTests.java:104]: Can probably remove that test actually since they are now tested in the methods above.

File: src/main/java/com/thebois/models/world/World.java [hunk 35]
```diff
@@ -50,4 +57,56 @@ public Object find() {
         return copy;
     }
 
+    /**
+     * Returns the structures in an arraylist as the interface IStructures.
+     *
+     * @return The list to be returned.
+     */
+    public ArrayList<IStructure> getStructures() {
+        final ArrayList<IStructure> copy = new ArrayList<>();
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/models/world/World.java:66]: Methods should return the most abstract type possible, so in this case you should try returning an `Iterable<IStructure>` (which just allows simple iteration in a foreach loop), and if the consumers of the method require more functionality, an `Collection<IStructure>`. If neither of those are enough, try `List<IStructure>` which should almost always work. Only return the concrete `ArrayList<IStructure>` when there is no other option.

In summary, try to use these in order from most abstract to least:
`Iterable`
`Collection`
`List`
`ArrayList`

File: src/main/java/com/thebois/models/world/World.java [hunk 36]
```diff
@@ -50,4 +57,56 @@ public Object find() {
...
+        for (final IStructure[] matrix : structuresMatrix) {
+            for (final IStructure structure : matrix) {
+                if (structure != null) {
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/models/world/World.java:69]: We maybe should've discussed this at the meeting today, but we should avoid null at all costs. Instead of using `null` to represent "no structure here", try using an `Option<IStructure>`. Or maybe creating an implementation of `IStructure`, `NoStructure` that is used just to represent empty tiles.

File: src/main/java/com/thebois/models/world/structures/House.java [hunk 37]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.thebois.models.Position;
+
+/**
+ * A structure of type House.
+ */
+public class House extends AbstractStructure {
+
+    /**
+     * Creates a house structure.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/models/world/structures/House.java:11]: Commenting constructors is difficult imo, but I've had success trying to describe *how* this constructor initializes the object rather than the fact that it does. In this case you could write "Creates a house structure at a given position in the world."

**Review by martinjonsson01 — APPROVED** [review]
Looks good! Ready for a merge. Exemplary tests. 

Left some comments with minor things you could change if you want to while you're waiting for 16-pawn-role-management to merge.

**Review by martinjonsson01 — COMMENTED** [review]
Looks great. I'll be submitting my PR very soon, I'm almost done.

**Comment by Matte02** [comment]
Done,
image
Very close to 100%, not sure how to get the last few percent.
image

and 
image

**Review by JacobBredin — APPROVED** [review]
Lgtm!

File: src/main/java/com/thebois/ColonyManagement.java [hunk 38]
```diff
@@ -59,12 +66,19 @@ public void create() {
...
-
+        // Model
+        createModels();
+        // Views
+        createGameView(tileSize);
+        createInfoView();
         // Screens
         gameScreen = new GameScreen(VIEWPORT_HEIGHT, VIEWPORT_WIDTH, uiSkin, gameView, infoView);
-
+        projector = gameScreen.getProjector();
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/ColonyManagement.java:76]: Feels out of place. either this line where its getting used. Or just call the getter when we need it (because I only see this being used once?)

File: src/main/java/com/thebois/controllers/StructureController.java [hunk 39]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Override
+    public boolean touchDown(final int x, final int y, final int pointer, final int button) {
+        if (button == LEFT_CLICK) {
+            final Vector2 worldSpaceCords = projector.unproject(x, y);
+            final float offSetX = gameWidget.getX();
+            final float offSetY = gameWidget.getY();
+            final int worldPosX = (int) ((worldSpaceCords.x - offSetX) / tileSize);
+            final int worldPosY = (int) ((worldSpaceCords.y - offSetY) / tileSize);
+            if (world.createStructure(worldPosX, worldPosY)) {
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/controllers/StructureController.java:53]: Remember we had discussion about shorter names even if its clear what they are, think we decided to write the whole thing. Just so that we all follow the same codestyle,

File: src/main/java/com/thebois/controllers/StructureController.java [hunk 40]
```diff
@@ -0,0 +1,68 @@
...
+                System.out.println("X: " + worldPosX + " Y: " + worldPosY);
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/controllers/StructureController.java:54]: Guess this was forgotten from some debugging?

File: src/main/java/com/thebois/models/world/World.java [hunk 41]
```diff
@@ -26,26 +31,58 @@
...
+        for (int y = 0; y < structureMatrix.length; y++) {
+            for (int x = 0; x < structureMatrix[y].length; x++) {
+                structureMatrix[y][x] = Optional.empty();
+            }
+        }
         initColony();
     }
 
+    /**
+     * Initiates the world with the given size.
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/world/World.java:53]: And something more right? 👀

**Review by jonathan-lindqvist — APPROVED** [review]
Great! 👍 

Just left some small comments, nothing big or super important. You can decide if a fix is needed or not.
